cpp/src/arrow/compute/kernels/vector_cumulative_sum.cc

```diff
@@ -0,0 +1,159 @@
+        uint8_t* out_bitmap = output->buffers[0]->mutable_data();
+        int64_t out_offset = input->offset();
+
+        if (input->data()->MayHaveNulls()) {
+          arrow::internal::CopyBitmap(input->null_bitmap_data(), input->offset(),
+                                      input->length(), out_bitmap, out_offset);
+          output->null_count = input->null_count();
+        } else {
+          bit_util::SetBitsTo(out_bitmap, out_offset, input->length(), true);
+          output->null_count = 0;
+        }
+
+        Sum(ctx->exec_context(), input, output, start);
+        return Status::OK();
+      case Datum::CHUNKED_ARRAY:
+        const auto& input = batch[0].chunked_array();
+
+        ArrayVector out_chunks;
+        for (const auto& chunk : input->chunks()) {
+          auto out_chunk = std::make_shared<ArrayData>(
+              chunk->type(), chunk->length(), chunk->null_count(), chunk->offset());
+
+          uint8_t* out_chunk_bitmap = out_chunk->buffers[0]->mutable_data();
+          if (chunk->data()->MayHaveNulls()) {
+            arrow::internal::CopyBitmap(chunk->null_bitmap_data(), chunk->offset(),
+                                        chunk->length(), out_chunk_bitmap,
+                                        out_chunk->offset());
+            out_chunk->null_count = chunk->null_count();
+          } else {
+            bit_util::SetBitsTo(out_chunk_bitmap, out_chunk->offset(), chunk->length(),
+                                true);
+            out_chunk->null_count = 0;
+          }
+
+          CType last_value = Sum(ctx->exec_context(), chunk, out_chunk, start);
+          start = last_value;
+          out_chunks.push_back(MakeArray(std::move(out_chunk)));
+        }
+
+        *out->chunked_array() = ChunkedArray(out_chunks, input->type());
+        return Status::OK();
+      default:
+        return Status::NotImplemented(
+            "Unsupported input type for function 'cumulative_sum': ",
+            batch[0].ToString());
+    }
+  }
+
+  static std::shared_ptr<KernelSignature> GetSignature(detail::GetTypeId get_id) {
+    return KernelSignature::Make({InputType::Array(get_id.id)}, OutputType(FirstType));
+  }
+};
+
+const FunctionDoc cumulative_sum_doc(
+    "Compute the cumulative sum over an array of numbers",
+    ("`values` must be an array of numeric type values.\n"
+     "`start` is a single value of the same type.\n"
+     "Return an array which is the cumulative sum computed over `values.`\n"
+     "Null entries remain in place but are not used in calucating sum.\n"
+     "`start` is an optional starting sum of computation."),
+    {"values", "start"});
+
+void RegisterVectorCumulativeSum(FunctionRegistry* registry) {
+  auto cumulative_sum = std::make_shared<VectorFunction>(
+      "cumulative_sum", Arity::Binary(), &cumulative_sum_doc);
+
+  auto add_kernel = [&](detail::GetTypeId get_id, ArrayKernelExec exec) {
+    VectorKernel kernel;
+    kernel.can_execute_chunkwise = true;
```

Comment:
Hmm, can this kernel work chunkwise? We'll have to make sure to test this

Reply:
Chunkwise here implies that chunks can be processed independently and in parallel, right?
I do not think cumsum can be processed chunkwise

Reply:
There was the idea of supporting cumsum in chunks with some state tracking. For example,
```python
start = 0
for batch in batches:
  auto cumulative_sum = CumulativeSum(batch, start);
  start = cumulative_sum[-1]
  SendDownstream(cumulative_sum)
```
cc @westonpace 

This would require higher level logic to know how to keep track of the `start` state and pass it on.
Should this be considered as `kernel.can_execute_chunkwise = false` and process all the data in a single batch?

Reply:
My preference would be to worry about chunkwise in the future, if that is a possibility.

This problem will need to be solved when we tackle window functions (cumulative sum is often expressed as a window function) so I expect we will know more once our work there has progressed.

Since the function is not scalar I don't think any of our execution engine code should allow it to be used today.  The only places we can plugin arbitrary functions right now are projection (err: not scalar) and aggregate (err: not aggregate).